File: Example/Cocoapod_UIKit/BrandPayObjcExample/ViewController.m
//
//  ViewController.m
//  BrandPayObjcExample
//
//  Created by 김진규 on 2023/12/13.
//  Copyright © 2023 com.tosspayments. All rights reserved.
//

#import "ViewController.h"
#import <WebKit/WebKit.h>
#import <BrandPayBase/BrandPayBase-Swift.h>
#import <OCRInterface/OCRInterface-Swift.h>
#import <BiometricInterface/BiometricInterface-Swift.h>

@interface ViewController () <WKNavigationDelegate, WebViewControllerType>

@property (nonatomic, strong) WKWebView *webView;

@end

@implementation ViewController

static NSString * const latestURL = @"LatestURL";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupWebView];
    [self setupRightBarButton];
    
    [self installAppBridges];
}

- (void)setupWebView {
    _webView = [[WKWebView alloc] initWithFrame:self.view.bounds];
    [self.view addSubview:_webView];
    _webView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [NSLayoutConstraint activateConstraints:@[
        [_webView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
        [_webView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
        [_webView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [_webView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor]
    ]];
    
    _webView.navigationDelegate = self;
#if DEBUG && __IPHONE_OS_VERSION_MAX_ALLOWED >= 160400
    [_webView setInspectable:true];
#endif
    
    NSString *urlString = [[NSUserDefaults standardUserDefaults] stringForKey:latestURL] ?: @"https://tosspayments.com";
    [_webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:urlString]]];
}

- (void)setupRightBarButton {
    self.navigationController.navigationBar.translucent = NO;
    UIBarButtonItem *item = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemEdit
                                                                          target:self
                                                                          action:@selector(didTapRightBarButton:)];
    self.navigationItem.rightBarButtonItem = item;
}

- (void)didTapRightBarButton:(id)sender {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Test URL을 입력하세요."
                                                                             message:@"클립보드의 내용이 URL일 경우 자동으로 붙여 넣습니다."
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        NSString *pasteBoardString = [UIPasteboard generalPasteboard].string;
        if (pasteBoardString && ![pasteBoardString isEqualToString:@""] && [NSURL URLWithString:pasteBoardString]) {
            textField.text = pasteBoardString;
        }
    }];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"취소"
                                                        style:UIAlertActionStyleCancel
                                                      handler:nil]];
    
    __weak typeof(self) weakSelf = self;
    [alertController addAction:[UIAlertAction actionWithTitle:@"확인"
                                                        style:UIAlertActionStyleDefault
                                                      handler:^(UIAlertAction * _Nonnull action) {
                                                          UITextField *urlTextField = alertController.textFields.firstObject;
                                                          NSString *urlString = urlTextField.text;
                                                          NSURL *url = [NSURL URLWithString:urlString];
                                                          if (url) {
                                                              [[NSUserDefaults standardUserDefaults] setObject:urlString forKey:latestURL];
                                                              [weakSelf.webView loadRequest:[NSURLRequest requestWithURL:url]];
                                                          }
                                                      }]];
    
    [self presentViewController:alertController animated:YES completion:nil];
}

- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [self showAlert:error.localizedDescription];
}

- (void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [self showAlert:error.localizedDescription];
}

- (void)showAlert:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"로드 실패"
                                                                             message:message
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"확인"
                                                        style:UIAlertActionStyleDefault
                                                      handler:nil]];
    [self presentViewController:alertController animated:YES completion:nil];
}

- (void)installAppBridges {
    WebScriptMessageHandler* messageHandler = [[WebScriptMessageHandler alloc] init];
    [messageHandler setController:self];
    
    [messageHandler registerWithAppBridge:[[GetAppInfoAppBridge alloc] init]];
    [messageHandler registerWithAppBridge:[[HasBiometricAuthAppBridge alloc] init]];
    [messageHandler registerWithAppBridge:[[GetBiometricAuthMethodsAppBridge alloc] init]];
    [messageHandler registerWithAppBridge:[[VerifyBiometricAuthAppBridge alloc] init]];
    [messageHandler registerWithAppBridge:[[RegisterBiometricAuthAppBridge alloc] init]];
    [messageHandler registerWithAppBridge:[[UnregisterBiometricAuthAppBridge alloc] init]];
    [_webView.configuration.userContentController addScriptMessageHandler:messageHandler name:@"ConnectPayAuth"];
    
//    * OCR 기능은 앱 패키지 별로 flk license file 로 관리됩니다.
    [messageHandler registerWithAppBridge:[[ScanOCRCardAppBridge alloc] initWithLicenseKeyFile:@"tosspayment_20220513.flk"]];
    [messageHandler registerWithAppBridge:[[IsOCRAvailableAppBridge alloc] init]];
    [_webView.configuration.userContentController addScriptMessageHandler:messageHandler name:@"ConnectPayOcr"];
    
    
    [_webView.configuration.userContentController addScriptMessageHandler:messageHandler name:@"connectWebViewAction"];
}

- (void)evaluateJavaScriptSafelyWithJavaScriptString:(NSString * _Nonnull)javaScriptString {
    [_webView evaluateJavaScript:javaScriptString completionHandler:^(id _Nullable result, NSError * _Nullable error) {
        
    }];
}

@end

File: Example/Cocoapod_UIKit/BrandPayWebExample/ViewController.swift
//
//  ViewController.swift
//  BrandPayWebInterface
//
//  Created by 김진규 on 2021/11/26.
//  Copyright © 2021 com.tosspayments. All rights reserved.
//

import WebKit
import BrandPayBase
import BiometricInterface
import OCRInterface

final class BrandPayWebInterfaceDemoViewController: UIViewController {
    enum Constants {
        static let latestURL: String = "LastestURL"
    }
    lazy var webView: WKWebView = {
        let webView = WKWebView()
        
        return webView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        setupWebView()
        setupRightBarButton()
        
        installAppBridges()
    }
    
    private func setupWebView() {
        view.addSubview(webView)
        webView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            webView.topAnchor.constraint(equalTo: view.topAnchor),
            webView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            webView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            webView.trailingAnchor.constraint(equalTo: view.trailingAnchor)
        ])
        webView.navigationDelegate = self
        
        let urlString = UserDefaults.standard.string(forKey: Constants.latestURL) ?? "https://tosspayments.com"
        webView.load(URLRequest(url: URL(string: urlString)!))
    }
    private func setupRightBarButton() {
        navigationController?.navigationBar.isTranslucent = false
        let item = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(didTapRightBarButton(_:)))
        navigationItem.rightBarButtonItem = item
    }
    
    @objc func didTapRightBarButton(_ sender: Any) {
        let alertController = UIAlertController(
            title: "Test URL을 입력하세요.",
            message: "클립보드의 내용이 URL일 경우 자동으로 붙여 넣습니다.",
            preferredStyle: .alert
        )
        alertController.addTextField(configurationHandler: { textField in
            if let pasteBoardString = UIPasteboard.general.string,
               !pasteBoardString.isEmpty, URL(string: pasteBoardString) != nil {
                textField.text = pasteBoardString
            }
        })
        alertController.addAction(
            UIAlertAction(title: "취소", style: .cancel, handler: nil)
        )
        alertController.addAction(
            UIAlertAction(title: "확인", style: .default, handler: { [weak self] _ in
                if let urlString = alertController.textFields?.first?.text,
                   let url = URL(string: urlString) {
                    UserDefaults.standard.set(urlString, forKey: Constants.latestURL)
                    self?.webView.load(URLRequest(url: url))
                }
            })
        )
        present(alertController, animated: true, completion: nil)
    }
}

extension BrandPayWebInterfaceDemoViewController: WebViewControllerType {
    func installAppBridges() {
        let messageHandler = WebScriptMessageHandler()
        messageHandler.controller = self
        
        messageHandler.register(appBridge: GetAppInfoAppBridge())
        messageHandler.register(appBridge: HasBiometricAuthAppBridge())
        messageHandler.register(appBridge: GetBiometricAuthMethodsAppBridge())
        messageHandler.register(appBridge: VerifyBiometricAuthAppBridge())
        messageHandler.register(appBridge: RegisterBiometricAuthAppBridge())
        messageHandler.register(appBridge: UnregisterBiometricAuthAppBridge())
        webView.configuration.userContentController.add(messageHandler, name: "ConnectPayAuth")
        
//         * OCR 기능은 앱 패키지 별로 flk license file 로 관리됩니다.
        messageHandler.register(appBridge: ScanOCRCardAppBridge(licenseKeyFile: "tosspayment_20220513.flk"))
        messageHandler.register(appBridge: IsOCRAvailableAppBridge())
        webView.configuration.userContentController.add(messageHandler, name: "ConnectPayOcr")
        
        // Deprecated: ConnectPayAuth, ConnectPayOcr 두개의 javascript object 로 나뉘어졌음
        webView.configuration.userContentController.add(messageHandler, name: "connectWebViewAction")
    }
    
    func evaluateJavaScriptSafely(javaScriptString: String) {
        webView.evaluateJavaScript(javaScriptString) { (_, _) in
            
        }
    }
    
    // SDK 내부에서 발생하는 에러를 받을 수 있는 콜백입니다.
    func onErrorOccurred(_ error: NSError) {
        // 1. NSError → BrandpayBiometricAuthError 변환
        guard let biometricAuthError = BrandpayBiometricAuthError.from(error) else {
            return
        }
        
        // 2. 확인할 수 있는 정보
        let message   = biometricAuthError.errorDescription    // 에러 메시지
        let traceId   = biometricAuthError.traceId             // 추적용 ID
        let underlying = biometricAuthError.underlyingError     // 발생한 원본 내부 에러
        
        print(String(describing: biometricAuthError))
    }
}

extension BrandPayWebInterfaceDemoViewController: WKNavigationDelegate {
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        showAlert(error.localizedDescription)
    }
    
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        showAlert(error.localizedDescription)
    }
    
    func showAlert(_ message: String) {
        let alertController = UIAlertController(title: "로드 실패", message: message, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "확인", style: .default, handler: nil))
        present(alertController, animated: true, completion: nil)
    }
}
